contracts/0.4.24/Lido.sol

```diff
@@ -999,10 +936,168 @@ contract Lido is ILido, StETH, AragonApp {
+
+        LimiterState.Data memory tokenRebaseLimiter = PositiveTokenRebaseLimiter.initLimiterState(
+            getMaxPositiveTokenRebase(),
+            _getTotalPooledEther(),
+            _getTotalShares()
+        );
+
+        tokenRebaseLimiter.applyCLBalanceUpdate(clBalanceDiff);
+        withdrawals = tokenRebaseLimiter.appendEther(_inputData.withdrawalVaultBalance);
+        elRewards = tokenRebaseLimiter.appendEther(_inputData.elRewardsVaultBalance);
+
+        // collect ETH from EL and Withdrawal vaults and send some to WithdrawalQueue if required
+        _processETHDistribution(
+            withdrawals,
+            elRewards,
+            _inputData.requestIdToFinalizeUpTo,
+            _inputData.finalizationShareRate
+        );
+
+        // distribute rewards to Lido and Node Operators
+        uint256 sharesMintedAsFees = _processRewards(clBalanceDiff, withdrawals, elRewards);
+        _applyCoverage(tokenRebaseLimiter);
+
+        (
+            postTotalPooledEther, postTotalShares
+        ) = _completeTokenRebase(
+            tokenRebaseLimiter, sharesMintedAsFees, _inputData.timeElapsed
+        );
+
+        emit ETHDistributed(
+            clBalanceDiff,
+            withdrawals,
+            elRewards,
+            _getBufferedEther()
+        );
+    }
+
+    function _completeTokenRebase(
+        LimiterState.Data memory _tokenRebaseLimiter,
+        uint256 _sharesMintedAsFees,
+        uint256 _timeElapsed
+    ) internal returns (uint256 postTotalPooledEther, uint256 postTotalShares) {
+        uint256 preTotalPooledEther = _tokenRebaseLimiter.totalPooledEther;
+        uint256 preTotalShares = _tokenRebaseLimiter.totalShares;
+
+        postTotalPooledEther = _getTotalPooledEther();
+        postTotalShares = _getTotalShares();
+
+        address postTokenRebaseReceiver = getLidoLocator().rebaseReceiver();
+        if (postTokenRebaseReceiver != address(0)) {
+            IPostTokenRebaseReceiver(postTokenRebaseReceiver).handlePostTokenRebase(
+                preTotalShares,
+                preTotalPooledEther,
+                postTotalShares,
+                postTotalPooledEther,
+                _sharesMintedAsFees,
+                _timeElapsed
+            );
+        }
+
+        emit TokenRebase(
+            preTotalShares,
+            preTotalPooledEther,
+            postTotalShares,
+            postTotalPooledEther,
+            _sharesMintedAsFees,
+            _timeElapsed
+        );
+    }
+
+    function _applyCoverage(LimiterState.Data memory _tokenRebaseLimiter) internal {
+        ISelfOwnedStETHBurner burner = ISelfOwnedStETHBurner(getLidoLocator().selfOwnedStEthBurner());
+        (uint256 coverShares, uint256 nonCoverShares) = burner.getSharesRequestedToBurn();
+        uint256 maxSharesToBurn = _tokenRebaseLimiter.deductShares(coverShares.add(nonCoverShares));
+
+        if (maxSharesToBurn > 0) {
+            burner.processLidoOracleReport(maxSharesToBurn);
```

Comment:
Maybe remove BURN_ROLE and burn in-place?

Reply:
done